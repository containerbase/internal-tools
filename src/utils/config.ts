import is from '@sindresorhus/is';
import { readFile } from '../util';
import type { ConfigFile } from './types';

const keys: (keyof ConfigFile)[] = [
  'datasource',
  'depName',
  'lookupName',
  'buildArg',
  'versioning',
  'latestVersion',
];

function checkArgs(
  cfg: ConfigFile,
  groups: Record<string, string | undefined>
): void {
  for (const key of keys) {
    if (!is.string(cfg[key]) && is.nonEmptyString(groups[key])) {
      cfg[key] = groups[key] as never;
    }
  }
}

export async function readDockerConfig(cfg: ConfigFile): Promise<void> {
  const dockerFileRe = new RegExp(
    '# renovate: datasource=(?<datasource>[a-z-]+?) depName=(?<depName>.+?)(?: lookupName=(?<lookupName>.+?))?(?: versioning=(?<versioning>[a-z-]+?))?\\s' +
      `(?:ENV|ARG) ${cfg.buildArg as string}=(?<latestVersion>.*)\\s`,
    'g'
  );
  const dockerfile = await readFile('Dockerfile');
  const m = dockerFileRe.exec(dockerfile);
  if (m && m.groups) {
    checkArgs(cfg, m.groups);
  }
}
