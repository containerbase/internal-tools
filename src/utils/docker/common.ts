import { exec } from '../../util';
import log from '../logger';
import type { ExecResult } from '../types';

export enum DockerPlatform {
  x86_64 = 'linux/amd64',
  aarch64 = 'linux/arm64',
}

export type DockerArch = 'x86_64' | 'aarch64';

export async function docker(...args: string[]): Promise<ExecResult> {
  return await exec('docker', [...args]);
}
export async function dockerRun(...args: string[]): Promise<void> {
  await docker('run', '--rm', ...args);
}

export async function dockerBuildx(...args: string[]): Promise<ExecResult> {
  return await docker('buildx', ...args);
}

type DockerTagConfig = {
  image: string;
  imagePrefix: string;
  src: string;
  tgt: string;
};

export async function dockerTag({
  image,
  imagePrefix,
  src,
  tgt,
}: DockerTagConfig): Promise<ExecResult> {
  return await exec('docker', [
    'tag',
    `${imagePrefix}/${image}:${src}`,
    `${imagePrefix}/${image}:${tgt}`,
  ]);
}

export async function dockerPrune(): Promise<void> {
  log('Pruning docker system');
  await docker('system', 'prune', '--force', '--all');
}

export async function dockerDf(): Promise<void> {
  log('Docker system disk usage');
  await docker('system', 'df');
}
