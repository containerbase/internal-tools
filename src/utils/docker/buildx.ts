import log from '../logger';
import { docker, dockerBuildx, dockerRun } from './common';

const SupportedPlatforms = 'arm64';

export async function init(): Promise<void> {
  const buildx = await dockerBuildx('ls');

  if (buildx.stdout.includes('renovatebot-builder')) {
    log('Buildx already initialized');
    return;
  }

  log.info('Configure buildx');

  await docker('info');
  // install emulations
  // https://github.com/docker/setup-qemu-action/blob/9d419fda7df46b2bcd38fadda3ec44f4748d25e1/src/main.ts#L22
  await dockerRun(
    '--privileged',
    'tonistiigi/binfmt',
    '--install',
    SupportedPlatforms
  );
  await dockerBuildx('version');

  await dockerBuildx(
    'create',
    '--name',
    'renovatebot-builder',
    '--driver',
    'docker-container',
    '--use'
  );

  await dockerBuildx('inspect', '--bootstrap');
}
