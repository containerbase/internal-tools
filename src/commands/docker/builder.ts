import { getInput, setFailed } from '@actions/core';
import is from '@sindresorhus/is';
import * as chalk from 'chalk';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { get as getVersioning } from 'renovate/dist/modules/versioning';
import { exec, getArg, isDryRun, readJson } from '../../util';
import { BuildsResult, addHostRule, getBuildList } from '../../utils/builds';
import { readDockerConfig } from '../../utils/config';
import { build, publish } from '../../utils/docker';
import { init } from '../../utils/docker/buildx';
import { dockerDf, dockerPrune, dockerTag } from '../../utils/docker/common';
import log from '../../utils/logger';
import type { Config, ConfigFile } from '../../utils/types';

export const MultiArgsSplitRe = /\s*(?:[;,]|$)\s*/;

function createTag(tagSuffix: string | undefined, version: string): string {
  return is.nonEmptyString(tagSuffix) && tagSuffix !== 'latest'
    ? `${version}-${tagSuffix}`
    : version;
}

async function buildAndPush(
  {
    imagePrefix,
    image,
    buildArg,
    buildArgs,
    buildOnly,
    cache,
    dryRun,
    tagSuffix,
    versioning,
    majorMinor,
    prune,
  }: Config,
  tobuild: BuildsResult
): Promise<void> {
  const builds: string[] = [];
  const failed: string[] = [];
  const ver = getVersioning(versioning);
  const versionsMap = new Map<string, string>();
  if (majorMinor) {
    for (const version of tobuild.versions) {
      const minor = ver.getMinor(version);
      const major = ver.getMajor(version);
      const isStable = ver.isStable(version);

      if (isStable && is.number(major) && `${major}` !== version) {
        versionsMap.set(`${major}`, version);
      }

      if (
        isStable &&
        is.number(major) &&
        is.number(minor) &&
        `${major}.${minor}` !== version
      ) {
        versionsMap.set(`${major}.${minor}`, version);
      }
    }
  }

  await exec('df', ['-h']);

  for (const version of tobuild.versions) {
    const tag = createTag(tagSuffix, version.replace(/\+.+/, ''));
    const imageVersion = `${imagePrefix}/${image}:${tag}`;
    log(`Building ${imageVersion}`);
    try {
      const minor = ver.getMinor(version);
      const major = ver.getMajor(version);
      const cacheTags: string[] = [tagSuffix ?? 'latest'];
      const tags: string[] = [];

      if (
        is.number(major) &&
        majorMinor &&
        versionsMap.get(`${major}`) === version
      ) {
        const nTag = createTag(tagSuffix, `${major}`);
        cacheTags.push(nTag);
        tags.push(nTag);
      }

      if (
        is.number(major) &&
        is.number(minor) &&
        majorMinor &&
        versionsMap.get(`${major}.${minor}`) === version
      ) {
        const nTag = createTag(tagSuffix, `${major}.${minor}`);
        cacheTags.push(nTag);
        tags.push(nTag);
      }

      if (version === tobuild.latestStable) {
        tags.push(tagSuffix ?? 'latest');
      }

      await build({
        image,
        imagePrefix,
        tag,
        cache,
        cacheTags,
        buildArgs: [...(buildArgs ?? []), `${buildArg}=${version}`],
        dryRun,
      });

      if (!buildOnly) {
        await publish({ image, imagePrefix, tag, dryRun });
        const source = tag;

        for (const tag of tags) {
          log(`Publish ${source} as ${tag}`);
          await dockerTag({ image, imagePrefix, src: source, tgt: tag });
          await publish({ image, imagePrefix, tag, dryRun });
        }
      }

      log(`Build ${imageVersion}`);
      builds.push(version);
    } catch (err) {
      log.error(err);
      failed.push(version);
    }

    await dockerDf();
    await exec('df', ['-h']);

    if (prune) {
      await dockerPrune();
      await exec('df', ['-h']);
    }
  }

  if (builds.length) {
    log('Build list:' + builds.join(' '));
  }

  if (failed.length) {
    log.warn('Failed list:' + failed.join(' '));
    throw new Error('failed');
  }
}

async function generateImages(config: Config): Promise<void> {
  const buildList = await getBuildList(config);

  if (!buildList?.versions.length) {
    setFailed(`No versions found.`);
    return;
  }

  await buildAndPush(config, buildList);
}

export async function run(): Promise<void> {
  const dryRun = isDryRun();
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  // TODO: validation
  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  const config: Config = {
    ...cfg,
    imagePrefix: getArg('image-prefix')?.replace(/\/$/, '') || 'renovate',
    image: cfg.image,
    depName: cfg.depName ?? cfg.image,
    buildArg: cfg.buildArg,
    buildArgs: getArg('build-args', { multi: true }),
    tagSuffix: getArg('tag-suffix') || undefined,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun,
    lastOnly: getInput('last-only') == 'true',
    buildOnly: getInput('build-only') == 'true',
    majorMinor: getArg('major-minor') !== 'false',
    prune: getArg('prune') === 'true',
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  };

  if (dryRun) {
    log('GitHub Actions branch detected - Force building latest, no push');
    config.lastOnly = true;
  }
  log('config:', JSON.stringify(config));

  const token = getArg('token');

  if (token) {
    addHostRule({ matchHost: 'github.com', token });
  }

  await init();

  await generateImages(config);

  log.info(chalk.blue('Processing done:', config.image));
}
