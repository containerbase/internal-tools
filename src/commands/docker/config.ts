import { init } from '../../utils/docker/buildx';

export async function run(): Promise<void> {
  await init();
}
