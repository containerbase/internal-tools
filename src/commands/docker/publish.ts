import { getInput } from '@actions/core';
import { isDryRun } from '../../util';
import { publish } from '../../utils/docker';
import log from '../../utils/logger';

export const MultiArgsSplitRe = /\s*(?:;|$)\s*/;
export async function run(): Promise<void> {
  const dryRun = isDryRun();
  let image = getInput('image', { required: true });
  let tags = getInput('tags')?.split(MultiArgsSplitRe).filter(Boolean);

  if (!tags?.length) {
    tags = ['latest'];
  }

  if (!image) {
    throw new Error('Missing image');
  }

  // istanbul ignore if
  if (image.startsWith('renovate/')) {
    log.warn('Image prefix can be removed:', image);
    image = image.substring(9);
  }

  for (const tag of tags) {
    await publish({ image, tag, dryRun });
  }
}
