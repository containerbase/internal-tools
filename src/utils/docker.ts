import is from '@sindresorhus/is';
import * as chalk from 'chalk';
import * as delay from 'delay';
import got, { HTTPError, Headers } from 'got';
import * as wwwAuthenticate from 'www-authenticate';
import { exists } from '../util';
import { cosign } from './cosign/common';
import { docker } from './docker/common';
import log from './logger';
import { ExecError } from './types';

export function splitImage(image: string): {
  registry: string;
  repository: string;
} {
  const parts = image.split('/');
  if (parts.length > 2) {
    return { registry: parts[0], repository: parts.slice(1).join('/') };
  } else {
    return { registry: 'index.docker.io', repository: image };
  }
}

export type DockerManifestV2 = {
  schemaVersion: number;
  config: {
    digest: string;
  };
};

export async function getAuthHeaders(
  registry: string,
  repository: string
): Promise<Headers> {
  try {
    const apiCheckUrl = `https://${registry}/v2/`;
    const apiCheckResponse = await got(apiCheckUrl, { throwHttpErrors: false });
    if (apiCheckResponse.headers['www-authenticate'] === undefined) {
      return {};
    }
    const authenticateHeader = new wwwAuthenticate.parsers.WWW_Authenticate(
      apiCheckResponse.headers['www-authenticate']
    );

    const authUrl = `${authenticateHeader.parms.realm}?service=${authenticateHeader.parms.service}&scope=repository:${repository}:pull`;
    const authResponse = (
      await got<{ token?: string; access_token?: string }>(authUrl, {
        responseType: 'json',
      })
    ).body;

    const token = authResponse.token || authResponse.access_token;
    if (!token) {
      throw new Error('Failed to obtain docker registry token');
    }
    return {
      authorization: `Bearer ${token}`,
    };
  } catch (err) {
    log.error(chalk.red('auth error'), (err as Error).message);
    throw new Error('Failed to obtain docker registry token');
  }
}

export enum DockerContentType {
  ManifestV1 = 'application/vnd.docker.distribution.manifest.v1+json',
  ManifestV1Signed = 'application/vnd.docker.distribution.manifest.v1+prettyjws',
  ManifestV2 = 'application/vnd.docker.distribution.manifest.v2+json',
}

const shaRe = /(sha256:[a-f0-9]{64})/;

export async function getRemoteImageId(
  registry: string,
  repository: string,
  tag = 'latest'
): Promise<string> {
  const headers = await getAuthHeaders(registry, repository);
  headers.accept = DockerContentType.ManifestV2;
  const url = `https://${registry}/v2/${repository}/manifests/${tag}`;

  try {
    const resp = await got<DockerManifestV2>(url, {
      headers,
      responseType: 'json',
    });

    switch (resp.headers['content-type']) {
      case DockerContentType.ManifestV2:
        return resp.body.config.digest;

      case DockerContentType.ManifestV1:
      case DockerContentType.ManifestV1Signed:
        // something wrong, we need to overwrite existing
        log.warn(
          chalk.yellow('Wrong response'),
          `Wrong response: ${resp.headers['content-type'] as string}`
        );
        return '<error>';

      default:
        throw new Error(
          `Unsupported response: ${resp.headers['content-type'] as string}`
        );
    }
  } catch (e) {
    if (e instanceof HTTPError && e.response.statusCode === 404) {
      // no image published yet
      return '<none>';
    }
    log.error(chalk.red('request error'), (e as Error).message);
    throw new Error('Could not find remote image id');
  }
}

export async function getLocalImageId(
  image: string,
  tag = 'latest'
): Promise<string> {
  const res = await docker('inspect', "--format='{{.Id}}'", `${image}:${tag}`);

  const [, id] = shaRe.exec(res.stdout) ?? [];

  if (!id) {
    log.error(res);
    throw new Error('Could not find local image id');
  }

  return id;
}

export type BuildOptions = {
  image: string;
  imagePrefix: string;
  cache?: string;
  cacheTags?: string[];
  tag?: string;
  dryRun?: boolean;
  buildArgs?: string[];
};

const errors = [
  'unexpected status: 400 Bad Request',
  ': no response',
  'error writing layer blob',
];

function canRetry(err: ExecError): boolean {
  return errors.some((str) => err.stderr.includes(str));
}

export async function build({
  image,
  imagePrefix,
  cache,
  cacheTags,
  tag = 'latest',
  dryRun,
  buildArgs,
}: BuildOptions): Promise<void> {
  const args = [
    'buildx',
    'build',
    '--load',
    `--tag=${imagePrefix}/${image}:${tag}`,
  ];

  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }

  if (is.string(cache)) {
    const cacheImage = `${imagePrefix}/${cache}:${image.replace(/\//g, '-')}`;
    args.push(`--cache-from=${cacheImage}-${tag}`);

    if (is.nonEmptyArray(cacheTags)) {
      for (const ctag of cacheTags) {
        args.push(`--cache-from=${cacheImage}-${ctag}`);
      }
    }

    if (!dryRun) {
      args.push(`--cache-to=type=registry,ref=${cacheImage}-${tag},mode=max`);
    }
  }

  for (let build = 0; ; build++) {
    try {
      await docker(...args, '.');
      break;
    } catch (e) {
      if (e instanceof ExecError && canRetry(e) && build < 2) {
        log.error(chalk.red(`docker build error on try ${build}`), e);
        await delay(5000);
        continue;
      }
      throw e;
    }
  }
}

type PublishOptions = {
  image: string;
  imagePrefix: string;
  tag: string;
  dryRun?: boolean;
};

export async function publish({
  image,
  imagePrefix,
  tag,
  dryRun,
}: PublishOptions): Promise<void> {
  const imageName = `${imagePrefix}/${image}`;
  const fullName = `${imageName}:${tag}`;
  log.info(chalk.blue('Processing image:'), chalk.yellow(fullName));

  log('Fetch new id');
  const newId = await getLocalImageId(imageName, tag);

  log('Fetch old id');
  const { registry, repository } = splitImage(imageName);
  const oldId = await getRemoteImageId(registry, repository, tag);

  if (oldId === newId) {
    log('Image uptodate, no push nessessary:', chalk.yellow(oldId));
    return;
  }

  log('Publish new image', `${oldId} => ${newId}`);
  if (dryRun) {
    log.warn(chalk.yellow('[DRY_RUN]'), chalk.blue('Would push:'), fullName);
  } else {
    await docker('push', fullName);
    await sign(fullName);
  }
  log.info(chalk.blue('Processing image finished:', newId));
}

async function sign(fullName: string): Promise<void> {
  if (!(await exists('cosign'))) {
    log.warn('Cosign is not installed. Skipping container signing');
    return;
  }

  log('Signing image', fullName);
  await cosign('sign', fullName);
}
