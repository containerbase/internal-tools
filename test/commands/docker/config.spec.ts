import { run } from '../../../src/commands/docker/config';

jest.mock('../../../src/utils/docker/buildx', () => ({
  init: () => Promise.resolve(),
}));

describe('commands/docker/config', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('works', async () => {
    await expect(run()).resolves.toBeUndefined();
  });
});
