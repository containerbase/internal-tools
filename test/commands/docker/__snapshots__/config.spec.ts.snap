// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Array [
  Array [
    "./bin/file.sh",
    Array [],
  ],
  Array [
    "docker",
    Array [
      "info",
    ],
  ],
  Array [
    "docker",
    Array [
      "buildx",
      "version",
    ],
  ],
  Array [
    "docker",
    Array [
      "run",
      "--rm",
      "--privileged",
      "multiarch/qemu-user-static",
      "--reset",
      "-p",
      "yes",
    ],
  ],
  Array [
    "docker",
    Array [
      "buildx",
      "create",
      "--name",
      "renovatebot-builder",
      "--driver",
      "docker-container",
      "--use",
    ],
  ],
  Array [
    "docker",
    Array [
      "buildx",
      "inspect",
      "--bootstrap",
    ],
  ],
]
`;
